File: 75-sort-colors/sort-colors.js
var swap = function(nums, i, j) {
    // nums[i] = nums[i] ^ nums[j];
    // nums[j] = nums[i] ^ nums[j];
    // nums[i] = nums[i] ^ nums[j];

    let temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
}

/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    //       low          mid                high
    // 0s         1s            unknown               2s

    let low = 0, mid = 0, high = nums.length - 1;
    while(mid <= high) {
        if(nums[mid] === 0) {
            swap(nums, mid, low);
            low++;
            mid++;
        } else if(nums[mid] === 1) {
            mid++;
        } else {
            swap(nums, mid, high);
            high--;
        }
    }
};

File: 75-sort-colors/sort-colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int n = nums.size();
        int zero = 0, one = 0, two = n-1;
        while(one <= two) {
            if(nums[one] == 0) {
                swap(nums[zero], nums[one]);
                zero++;
                one++;
            } else if(nums[one] == 1) {
                one++;
            } else {
                swap(nums[one], nums[two]);
                two--;
            }
        }
    }
};